### PR TITLE
HDDS-3465. OM Failover retry happens too quickly when new leader suggested and retrying on same OM.

### PR DESCRIPTION
**Pull request by umamaheswararao**
## What changes were proposed in this pull request?

When leader was suggested, we are updating lastAttemptedOMid correctly, so that waitTime can be calculated by incrementing if failover to same OM.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-3465

## How was this patch tested?
Added 2 tests which reproduce the case and after fix, they both passed.

### CODE REVIEW

File: hadoop-ozone/common/src/test/java/org/apache/hadoop/ozone/om/ha/TestOMFailoverProxyProvider.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+  private final static String DUMMY_NODE_ADDR = "0.0.0.0:8080";
+
+  /**
+   * 1. Create FailoverProvider with 3 OMs.
+   * 2. Try failover to different node: WaitTime should be 0.
+   * 3. Try failover to same node: wiatTime should increment
+   * attempts*waitBetweenRetries
+   */
+  @Test
+  public void testWaitTimeWithSameNodeFailover() throws IOException {
```
[the remark this answers is not in the document]
Comment (umamaheswararao) [reply hadoop-ozone/common/src/test/java/org/apache/hadoop/ozone/om/ha/TestOMFailoverProxyProvider.java]: @adoroszlai , Thank you for reviews. I have refactored tests and added few more cases. Please take a look at it.

File: hadoop-ozone/common/src/test/java/org/apache/hadoop/ozone/om/ha/TestOMFailoverProxyProvider.java [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+  private final static String OM_SERVICE_ID = "om-service-test1";
+  private final static String NODE_ID_BASE_STR = "omNode-";
+  private final static String DUMMY_NODE_ADDR = "0.0.0.0:8080";
+  private OMFailoverProxyProvider provider;
+  private long waitBetweenRetries;
+
+  @Before
+  public void init() throws Exception {
+    OzoneConfiguration config = new OzoneConfiguration();
+    long numNodes = 3;
```
[the remark this answers is not in the document]
Comment (umamaheswararao) [reply hadoop-ozone/common/src/test/java/org/apache/hadoop/ozone/om/ha/TestOMFailoverProxyProvider.java]: Hay Attila, Thanks for reviews. It make sense to make it member variable. I remember I extracted it for the purpose and lost that. BTW, that variable should be int, not long. Just updated that too.
Take a look.
And the pr test failures seems unrelated, I had a quick look.

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/ha/OMFailoverProxyProvider.java [hunk 3]
```diff
@@ -318,9 +318,12 @@ synchronized boolean updateLeaderOMNodeId(String newLeaderOMNodeId) {
     if (!currentProxyOMNodeId.equals(newLeaderOMNodeId)) {
       if (omProxies.containsKey(newLeaderOMNodeId)) {
         currentProxyOMNodeId = newLeaderOMNodeId;
+        lastAttemptedOM = currentProxyOMNodeId;
```
Comment (hanishakoneru) [inline hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/ha/OMFailoverProxyProvider.java:321]: We are updating lastAttemptedOM to the new OM which will be tried after failover. The assignment should happen before the currentProxyOMNodeId is updated with newLeaderOMNodeId.

Comment (umamaheswararao) [reply hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/ha/OMFailoverProxyProvider.java:320]: @hanishakoneru good catch. I added new test to cover this case too. Please take a look at it.

**Comment by hanishakoneru** [comment]
Thanks Uma for working on this. 
LGTM. +1.

**Pull request merged**